cancermuts/datasources.py

```diff
@@ -1161,6 +1165,159 @@ def _get_revel_from_gc(self, mutation, gc):
+    @logger_init
+    def __init__(self, database_dir, database_files=None):
+        description = "Local REVEL score annotation using Ensembl transcript ID"
+        super(RevelLocal, self).__init__(name='RevelLocal', version='1.0', description=description)
+
+        self._database_dir = database_dir
+
+        if database_files is None:
+            self._database_files = {
+                'revel': os.path.join(self._database_dir, 'revel_with_transcript_ids')
+            }
+        else:
+            self._database_files = database_files
+
+        self._supported_metadata = {'revel_score': self._get_revel}
+
+    def add_metadata(self, sequence, md_type=['revel_score']):
+        if not sequence.uniprot_ac:
+            self.log.error("Sequence is missing UniProt accession; skipping.")
+            return
+
+        if type(md_type) is str:
+            md_types = [md_type]
+        else:
+            md_types = md_type
+
+        metadata_functions = []
+        for md in md_types:
+            try:
+                metadata_functions.append(self._supported_metadata[md])
+            except KeyError:
+                self.log.warning("RevelLocal doesn't support metadata type %s" % md)
+
+        try:
+            transcript_id = self._get_transcript_id(sequence)
+        except Exception as e:
+            self.log.error(f"Failed to retrieve transcript ID for {sequence}: {e}")
+            return
+
+        for pos in sequence.positions:
+            for mut in pos.mutations:
+                for add_md in metadata_functions:
+                    add_md(mut, transcript_id)
+
+    def _get_transcript_id(self, sequence):
+
+        if sequence.is_canonical:
+            this_upac = sequence.uniprot_ac
+        else:
+            if not sequence.isoform:
+                raise UnexpectedIsoformError("Sequence is non-canonical but has no isoform ID.")
+            this_upac = sequence.isoform
+
+        url = f"https://rest.uniprot.org/uniprotkb/{this_upac}.json"
+        self.log.debug(f"Querying UniProt for {this_upac} at {url}")
+        response = rq.get(url)
+
+        if response.status_code != 200:
+            raise ConnectionError(f"Failed to fetch UniProt JSON for {this_upac}: status {response.status_code}")
+
+        data = response.json()
+
+        xrefs = data.get('uniProtKBCrossReferences', [])
+        for xref in xrefs:
+            if xref.get('database') == 'Ensembl':
+                transcript_id = xref.get('id')
+                if transcript_id:
+                    transcript_id = transcript_id.split('.')[0]
+                    print(f"[REVEL] Transcript ID for {this_upac}: {transcript_id}")
+                    self.log.info(f"Using Ensembl transcript ID: {transcript_id} for {this_upac}")
+                    return transcript_id
+
+        raise KeyError(f"No Ensembl transcript ID found in UniProt entry for {this_upac}")
+
+    def _get_revel(self, mutation, transcript_id):
```

Comment:
it's a bit wasteful that we parse the REVEL file for every mutation - instead we should parse it once, and then keep the mutations that we need to keep. Notice that you can change the structure of `add_metadata` somewhat to make this easier